btf/core.go

```diff
@@ -1040,13 +1045,293 @@ func coreAreMembersCompatible(localType Type, targetType Type) error {
+	if level <= 0 {
+		return fmt.Errorf("types are nested too deep")
+	}
+
+	var (
+		localTs, targetTs typeDeque
+		l, t              = &localType, &targetType
+		depth             = 0
+	)
+
+	for ; l != nil && t != nil; l, t = localTs.Shift(), targetTs.Shift() {
+		if depth >= maxResolveDepth {
+			return errors.New("types are nested too deep")
+		}
+
+		localType = UnderlyingType(*l)
+		targetType = UnderlyingType(*t)
+
+		if !coreNamesMatch(localType.TypeName(), targetType.TypeName()) {
+			return fmt.Errorf("type name %q don't match %q: %w", localType.TypeName(), targetType.TypeName(), errIncompatibleTypes)
+		}
+
+		switch lv := (localType).(type) {
+		case *Void:
+			if _, ok := targetType.(*Void); !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+		case *Fwd:
+			if behindPtr {
+				if tv, ok := targetType.(*Fwd); ok {
+					if lv.Kind != tv.Kind {
+						return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+					}
+
+					continue
+				}
+
+				if _, ok := targetType.(*Struct); ok && lv.Kind == FwdStruct {
+					continue
+				}
+
+				if _, ok := targetType.(*Union); ok && lv.Kind == FwdUnion {
+					continue
+				}
+
+				return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if tv, ok := targetType.(*Fwd); ok && tv.Kind == lv.Kind {
+				continue
+			}
+
+			return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+
+		case *Enum:
+			tv, ok := targetType.(*Enum)
+			if !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if err := coreEnumsMatch(lv, tv); err != nil {
+				return err
+			}
+
+		case *Struct, *Union:
+			if behindPtr {
+				if reflect.TypeOf(localType) == reflect.TypeOf(targetType) {
+					continue
+				}
+
+				tv, ok := targetType.(*Fwd)
+				if !ok {
+					return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+				}
+
+				if _, ok := lv.(*Struct); ok && tv.Kind == FwdStruct {
+					continue
+				}
+
+				if _, ok := lv.(*Union); ok && tv.Kind == FwdUnion {
+					continue
+				}
+
+				return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if err := coreCompositesMatch(localType, targetType, behindPtr, level); err != nil {
+				return err
+			}
+
+		case *Int:
+			tv, ok := targetType.(*Int)
+			if !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if lv.Size != tv.Size || lv.Encoding&Signed != tv.Encoding&Signed {
```

Comment:
So in the C version if this function they also do a binary AND https://github.com/libbpf/libbpf/blob/master/src/relo_core.c#L1629

This confused me as well since the `IntEncoding` constants have this `These may look like they are flags, but they aren't.` on them. But on the other hand we always take the value for `IntEncoding` directly from the binary BTF format.

So clearly, we assume its an enum. And at least the writer of this CO-RE function assumes its flags. So its likely someone is wrong. But if you reason through it, we have 0, 1, 2, 4. So if it is an enum the current code works, since only `Signed & Signed` would produce anything other than 0. And if its flags then we have the same behavior as libbpf. So I thought this was a good way of hedging my bets without having to figure out who is wrong :grin:.

Reply:
I've sent an email about this to bpf@ at some point, so I'd go with the "this is an enum" interpretation. Maybe there is a link in the commit message for the commit that introduced the comment you referenced?

Reply:
Right. I found the commit + issue + email. So then I am assuming the author of this feature in libbpf was also confused and incorrect in this case. 

Will switch this to using it as an enum!